src/Features/Core/Portable/QuickInfo/CommonSemanticQuickInfoProvider.cs

```diff
@@ -151,85 +169,84 @@ internal abstract partial class AbstractSemanticQuickInfoProvider : AbstractQuic
             CancellationToken cancellationToken)
         {
             var descriptionService = workspace.Services.GetLanguageServices(token.Language).GetService<ISymbolDisplayService>();
+            var formatter = workspace.Services.GetLanguageServices(semanticModel.Language).GetService<IDocumentationCommentFormattingService>();
+            var syntaxFactsService = workspace.Services.GetLanguageServices(semanticModel.Language).GetService<ISyntaxFactsService>();
+            var showWarningGlyph = supportedPlatforms != null && supportedPlatforms.HasValidAndInvalidProjects();
+            var showSymbolGlyph = true;
 
-            var sections = await descriptionService.ToDescriptionGroupsAsync(workspace, semanticModel, token.SpanStart, symbols.AsImmutable(), cancellationToken).ConfigureAwait(false);
+            var groups = await descriptionService.ToDescriptionGroupsAsync(workspace, semanticModel, token.SpanStart, symbols.AsImmutable(), cancellationToken).ConfigureAwait(false);
+            var sections = new List<QuickInfoSection>(groups.Count);
+
+            if (groups.ContainsKey(SymbolDescriptionGroups.MainDescription) && !groups[SymbolDescriptionGroups.MainDescription].IsDefaultOrEmpty)
+            {
+                sections.Add(QuickInfoSection.Create(QuickInfoSectionKinds.Description, groups[SymbolDescriptionGroups.MainDescription]));
+            }
 
+            var documentationContent = GetDocumentationContent(symbols, groups, semanticModel, token, formatter, syntaxFactsService, cancellationToken);
+            if (syntaxFactsService.IsAwaitKeyword(token) &&
+                (symbols.First() as INamedTypeSymbol)?.SpecialType == SpecialType.System_Void)
+            {
+                documentationContent = default;
+                showSymbolGlyph = false;
+            }
 
-            var mainDescriptionBuilder = new List<TaggedText>();
-            if (sections.TryGetValue(SymbolDescriptionGroups.MainDescription, out var parts))
+            if (!documentationContent.IsDefaultOrEmpty)
             {
-                mainDescriptionBuilder.AddRange(parts);
+                sections.Add(QuickInfoSection.Create(QuickInfoSectionKinds.DocumentationComments, documentationContent));
             }
 
-            var typeParameterMapBuilder = new List<TaggedText>();
-            if (sections.TryGetValue(SymbolDescriptionGroups.TypeParameterMap, out parts))
+            if (groups.ContainsKey(SymbolDescriptionGroups.TypeParameterMap) && !groups[SymbolDescriptionGroups.TypeParameterMap].IsDefaultOrEmpty)
```

Comment:
Now that we have local functions, any way you can simplify the duplication here as we're doing this for different groups?

Reply:
Good idea.